### ISSUE TITLE
Issue with calculating elasticity matrix

### ISSUE THREAD

**Issue by kirichoi** [sys-bio/roadrunner#428]
When using the same model as reference in https://github.com/sys-bio/roadrunner/issues/420, elasticity matrix is incorrect.  However, reaction fluxes and species concentration are good and manual approximation results in correct value, so this must be something to do with the way we are calculating elasticity matrix.

**Comment by kirichoi** [comment]
Checking uEE, values are very off.  The routine uses `getFloatingSpeciesConcentrations()` to get the species concentration which differs from the actual value for some reason.  The real issue comes from when calculating the reaction rates. I attempted to implement practically the same functional code but the output is different. I don't understand why we are not using `getValue()` instead, which seems to work safely. Are there any reasons?

**Comment by hsauro** [comment]
I just ran the following using a very simple model, the elasticity
estimation routines seem to be very broken, estimates are way off:

# -*- coding: utf-8 -*-
"""
Created on Fri Nov 17 19:39:54 2017

@author: Herbert Sauro
"""

import tellurium as te
import roadrunner

r = te.loada("""
     J1: $Xo -> S1; (Vm1/Km1)*( Xo - S1 /Keq1)/(1+ Xo /Km1+ S1 /Km2);
     J2: S1 -> S2;  (Vm2/Km3)*( S1 - S2 /Keq2)/(1+ S1 /Km3+ S2 /Km4);
     J3: S2 -> $X1; Vm3*S2 /(Km5 +  S2 );

     Xo = 10;
     Keq1 = 0.56;
     Keq2 = 0.38;
     Keq3 = 0.76;

     Vm1 = 4.5; Vm2 = 3.4; Vm3 = 2.1;

     Km1 = 0.55; Km2 = 0.65;
     Km3 = 0.78; Km4 = 1.3;
     Km5 = 0.8;
""")

print 'True values for unscaled elasticties = 0.461589 and -1.89898'
print 'Unscaled Elasticities via roadrunner: ', r.getuEE ('J2', 'S1'), ', '
, r.getuEE ('J2', 'S2')
print 'Scaled Elasticities via roadrunner: ', r.getEE ('J2', 'S1'), ', ' ,
r.getEE ('J2', 'S2')

r.steadyState();

refJ2 = r.J2;
refS1 = r.S1;
refS2 = r.S2;

print 'Numerical estimates by simple two point estimation:'
r.S1 = 1.025*r.S1;
print "Numerical estimate: ue1 = ", (refJ2 - r.J2)/(refS1 - r.S1);

r.S1 = refS1;
r.S2 = 1.025*r.S2;

print "Numerical estimate: ue2 = ", (refJ2 - r.J2)/(refS2 - r.S2);

**Comment by hsauro** [comment]
getValue is not an efficient method as it has to look up values via a
string and it has to check what type it is.

Herbert

**Comment by hsauro** [comment]
Here is the output from the script I sent earlier. The estimates are an
order of magnitiude out, I don't think I made a mistake. My rough numerical
estimate match Copasi's estimates.  I am sure the elasticity estimation
used to work. Don't our regression tests check the elasicity functions? If
not we should have another one day hackthon to write regression tests. Are
regression tests run when roadrunner is built?

True values for unscaled elasticties = 0.461589 and -1.89898
Unscaled Elasticities via roadrunner:  4.35867379856 ,  -11.4708840001
Scaled Elasticities via roadrunner:  0.0 ,  -0.0
Numerical estimates by simple two point estimation:
Numerical estimate: ue1 =  0.453131256293
Numerical estimate: ue2 =  -1.89430950765

**Comment by kirichoi** [comment]
from @hsauro:

Here is example where the unscaled elasticity is completely wrong. Model uses first-order irreversible kinetcs so that the unscaled elasticity is just the value of the rate constant, that is 0.24. 

Herbert
```
import tellurium as te
import roadrunner

r = te.loada("""
     $Xo -> S1;  k1*Xo;
      J2: S1 -> S2;  k2*S1;
      S2 -> $X1; k3*S2;
     
      k1 = 0.1;
      k2 = 0.24;
      k3 = 0.14;
      Xo = 10
""")

r.steadyState()

print 'Unscaled: ', r.getuEE ('J2', 'k2'), ' Should be 0.24'
print 'Scaled: ', r.getEE ('J2', 'k2'), ' This is actually correct'

refJ = r.J2
refS1 = r.S1;
r.S1 = 1.05*r.S1;

print 'Numerical Estimate:', (refJ - r.J2)/(refS1 - r.S1)
```

Output:

Unscaled:  4.16666666667  Should be 0.24
Scaled:  1.0  This is actually correct
Numerical Estimate: 0.24

**Comment by kirichoi** [comment]
Update on this issue:

I noticed that our way of getting species concentration and reaction rate is correct and the same as COPASI. However, the unscaled elasticity matrix out of these values is different from COPASI, The funny thing is, when I try to obtain species concentration through `getValue()` I get completely different value from COPASI, but using it through the same calculation as before results in the same unscaled elasticity as COPASI. 

Now, digging deeper, I realized that our species amounts matches perfectly which what I get out of `getValue()` and probably COPASI is using amounts by default for calculating elasticity matrices, just like @0u812 suggested. I am not sure when `getValue()` defaulted into using amounts though. **However, `getuEE()` is hard-coded into using concentrations, which was causing this issue.** Everything works fine when calling `getUnscaledElasticityMatrix()` while using Kyle's suggestion:

`roadrunner.Config.setValue(roadrunner.Config.ROADRUNNER_JACOBIAN_MODE, roadrunner.Config.ROADRUNNER_JACOBIAN_MODE_AMOUNTS)`

`getUnscaledElasticityMatrix()` respects this flag while `getuEE()` does not.

The above model by @hsauro is comparing two different things. It should call `getuEE('J2', 'S1')` not `getuEE('J2', 'k2'). Then the values are correct, and concentration in this model is identical to amount so above flag is not necessary. 

Basically, we are good for elasticity, nothing low level is broken, we just need to redefine what to use when calculating elasticity. I will not close this until I update the code so that 'getuEE()' respects the flag.

**Comment by hsauro** [comment]
Note that elasticities are defined with respect to concentations, not
amounts. There is no elasticity defined in the literature that computes at
the partial derivatives of the reaction rate with respect to changes in
amounts.

Herbert

**Comment by kirichoi** [comment]
There is a possibility that COPASI is the problem here. As I said, our concentration is the same as COPASI. Our flux is also identical to COPASI. Under this condition, numerical estimate results in the same value as us.

**Comment by hsauro** [comment]
So are you saying that our elasticity calculations are correct? If so I
don't agree, given the example I posted our calculatins are so badly off.

Herbert

**Comment by kirichoi** [comment]
Both examples actually work. For the first example, you need to call `getuEE()` after steadystate because it attempts to calculate elasticity at t = 0 otherwise. If you do so, the output is:

```
True values for unscaled elasticties = 0.461589 and -1.89898
Unscaled Elasticities via roadrunner:  0.461589051295 ,  -1.8989790208
Scaled Elasticities via roadrunner:  1.62827214177 ,  -1.47381396832
Numerical estimates by simple two point estimation:
Numerical estimate: ue1 =  0.453131256293
Numerical estimate: ue2 =  -1.89430950765
```

The second example is comparing two different things. It should compare `r.getuEE ('J2', 'S1')` not `r.getuEE ('J2', 'k2')`. If you do so, the output is:

```
Unscaled:  0.24  Should be 0.24
Scaled:  1.0  This is actually correct
Numerical Estimate: 0.24
```

**Comment by hsauro** [comment]
However the reason I did r.getuEE ('J2', 'k1') is that I know the uee must
be k1 so its easy to check.

We still have the problem though that we can't compute the steady state of
the Keonig model.

Herbert

**Comment by hsauro** [comment]
Yes you're right, the uee (j2, k2) should be S1 and S1 is correctly computed.  Question remains however why can't we compute the steady state for the Koenig model and why are some of the jacobian entries wrong?

**Comment by kirichoi** [comment]
`getuEE()` needs some improvement (e.g. computing steady-state automatically) and that's why I did not closed this issue. As for the Koenig model, I believe you suggested the reason why we can't calculate steady-state is because the Jacobian is not invertible  (with rows and columns of zeros)? 
All Jacobian related issue will should be discussed in https://github.com/sys-bio/roadrunner/issues/420. I think the reason why our Jacobian entries are different is because our list of dependent species is different and that's causing issues with the link matrix.

**Comment by hsauro** [comment]
You shouldn't compute the steady state automatically when calling uEE
because we also need to compute elasticities under non-steady state
conditions. Note that the elasticity is not a steady state property, its
just dv/ds where s could have any value, not necessarily the steady state.

Herbert

**Comment by kirichoi** [comment]
I see. I will keep it as is.
